### PR TITLE
Add Hikari Pool Name tag 

### PR DESCRIPTION
**Pull request by jordan-wong**
# What Does This Do
Adds instrumentation for Hikari Pool Name. If using a Hikari pool, adds a `db.pool.name` tag to span.

# Motivation
Helps with issue: https://github.com/DataDog/dd-trace-java/issues/7152

# Additional Notes

# Contributor Checklist

- [ ] Format the title [according the contribution guidelines](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#title-format)
- [ ] Assign the `type:` and (`comp:` or `inst:`) labels in addition to [any usefull labels](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#labels)
- [ ] Squash your commits prior merging or merge using GitHub's [Squash and merge](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#squash-and-merge-your-commits)
- [ ] Don't use `close`, `fix` or any [linking keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) when referencing an issue.  
  Use `solves` instead, and assign the PR [milestone](https://github.com/DataDog/dd-trace-java/milestones) to the issue
- [ ] Update the [public documentation](https://docs.datadoghq.com/tracing/trace_collection/library_config/java/) in case of new configuration flag or behavior

Jira ticket: [PROJ-IDENT]

### CODE REVIEW

**Comment by jordan-wong** [comment]
mistake ping, messed up rebase

File: dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/DataSourceInstrumentation.java [hunk 1]
```diff
@@ -61,12 +62,17 @@ public static AgentScope start(@Advice.This final DataSource ds) {
         // Don't want to generate a new top-level span
         return null;
       }
-
       final AgentSpan span = startSpan(DATABASE_CONNECTION);
       DECORATE.afterStart(span);
 
       span.setResourceName(DECORATE.spanNameForMethod(ds.getClass(), "getConnection"));
 
+      if (ds instanceof HikariDataSource) {
```
Comment (amarziali) [inline dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/DataSourceInstrumentation.java:70]: You cannot assume that the customer has this dependency on the classpath. Making that instrumentation depending to hikari jars will make that unusable for people not having it because muzzle blocks this when not all the classes needed are available at runtime (reason why the muzzle task is failing on the CI).
I strongly suggest to do a separate instrumentation for this (i.e. `HikariDataSourceInstrumentation`) instrumenting `HikariDataSource` only. You should also have a separate `muzzleDirective` for it if you want to test muzzle correctly

File: dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/DataSourceInstrumentation.java [hunk 2]
```diff
@@ -61,7 +61,6 @@ public static AgentScope start(@Advice.This final DataSource ds) {
         // Don't want to generate a new top-level span
         return null;
       }
-
```
Comment (nayeem-kamal) [inline dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/DataSourceInstrumentation.java:64]: We can revert this line back for cleanliness

File: dd-java-agent/instrumentation/jdbc/src/test/groovy/RemoteJDBCInstrumentationTest.groovy [hunk 3]
```diff
@@ -183,7 +183,7 @@ abstract class RemoteJDBCInstrumentationTest extends VersionedNamingTestBase {
     mysql.start()
     PortUtils.waitForPortToOpen(mysql.getHost(), mysql.getMappedPort(MySQLContainer.MYSQL_PORT), 5, TimeUnit.SECONDS)
     jdbcUrls.put(MYSQL, "${mysql.getJdbcUrl()}")
-    sqlserver = new MSSQLServerContainer().acceptLicense().withPassword(jdbcPasswords.get(SQLSERVER))
+    sqlserver = new MSSQLServerContainer(MSSQLServerContainer.IMAGE).acceptLicense().withPassword(jdbcPasswords.get(SQLSERVER))
```
Comment (nayeem-kamal) [inline dd-java-agent/instrumentation/jdbc/src/test/groovy/RemoteJDBCInstrumentationTest.groovy:186]: I think we can revert the changes to this file since it seems unrelated to your testing

File: dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/HikariDataSourceInstrumentation.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  public void methodAdvice(MethodTransformer transformer) {
+    transformer.applyAdvice(
+        named("getConnection"),
+        HikariDataSourceInstrumentation.class.getName() + "$HikariGetConnectionAdvice");
+  }
+
+  public static class HikariGetConnectionAdvice {
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static void start(@Advice.This final HikariDataSource ds) {
+      if (activeSpan() == null) {
```
Comment (amarziali) [inline dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/HikariDataSourceInstrumentation.java:36]: This won't work because it relies on the `DatasourceInstrumentation` that's disabled by default. The test passes because `injectSysConfig("dd.integration.jdbc-datasource.enabled", "true")`
Also there are two advices kickin in on the same method enter and here we assume an execution order.
Also, in order to do a split-by-tag on the `jdbc.query` spans you should add this tag on the statment span and not on the connection span.

An approach might be here to store that pool information on the `DBInfo` state that's stored when the connection is created. In fact `DriverInstrumentation` already link 'java.sql.Connection` to `DbInfo`. So here you can use the `InstrumentationContext` to retrieve that dbInfo value and enrich it with the hikari pool. Then, you shall ensure that that the tag is added when the statement spans are created

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/jdbc/DBInfo.java [hunk 5]
```diff
@@ -141,6 +147,11 @@ public Builder port(Integer port) {
       return this;
     }
 
+    public Builder hikariPoolName(String poolName) {
```
Comment (amarziali) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/jdbc/DBInfo.java:150]: I would rather use a generic name (i.e. `poolName`) not linked to `hikari` since we might add this also for other pool implementations in the future

File: internal-api/src/main/java/datadog/trace/bootstrap/instrumentation/api/Tags.java [hunk 6]
```diff
@@ -47,6 +47,7 @@ public class Tags {
   public static final String DB_HOST = "db.host";
   public static final String DB_SCHEMA = "db.schema";
   public static final String MESSAGE_BUS_DESTINATION = "message_bus.destination";
+  public static final String HIKARI_POOL_NAME = "hikari.poolname";
```
Comment (amarziali) [inline internal-api/src/main/java/datadog/trace/bootstrap/instrumentation/api/Tags.java:50]: may we use generic names like `db.pool.name`?

File: dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/HikariDataSourceInstrumentation.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+      try {
+        if (con.isWrapperFor(Connection.class)) {
+          unwrapped = con.unwrap(Connection.class);
+        }
+      } catch (Throwable t) {
+        return;
+      }
+
+      String hikariPoolname = ds.getPoolName();
+      DBInfo dbInfo = InstrumentationContext.get(Connection.class, DBInfo.class).get(unwrapped);
```
Comment (amarziali) [inline dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/HikariDataSourceInstrumentation.java:67]: nit: check for null before using `get()`

**Review by amarziali — APPROVED** [review]
LGTM. Hikari 6 is not yet tested hence the lockfile should be bumped before merging

**Review by nayeem-kamal — APPROVED** [review]
LGTM

File: internal-api/src/main/java/datadog/trace/bootstrap/instrumentation/api/Tags.java [hunk 8]
```diff
@@ -47,6 +47,7 @@ public class Tags {
   public static final String DB_HOST = "db.host";
   public static final String DB_SCHEMA = "db.schema";
   public static final String MESSAGE_BUS_DESTINATION = "message_bus.destination";
+  public static final String DB_POOL_NAME = "db.pool.name";
```
Comment (bouwkast) [inline internal-api/src/main/java/datadog/trace/bootstrap/instrumentation/api/Tags.java:50]: Just to check :)

The PR description is outdated I think as it states that it'd add `hikari.poolname` as the tag, but it appears to be `db.pool.name`. Might be worth to update the PR description
